emu_base/pulser_adapter.py

```diff
@@ -77,13 +100,83 @@ def _xy_interaction(sequence: pulser.Sequence) -> torch.Tensor:
     return interaction_matrix
 
 
+def _get_amp_factors(
+    samples: pulser.sampler.SequenceSamples,
+    amp_sigma: float,
+    laser_waist: float | None,
+    qubit_positions: list[torch.Tensor],
+    q_ids: tuple[str, ...],
+) -> dict[int, torch.Tensor]:
+    def perp_dist(pos: torch.Tensor, axis: torch.Tensor) -> Any:
+        return torch.linalg.vector_norm(pos - torch.vdot(pos, axis) * axis)
+
+    times_to_amp_factors: dict[int, torch.Tensor] = {}
+    for ch, ch_samples in samples.channel_samples.items():
+        ch_obj = samples._ch_objs[ch]
+        prop_dir = torch.tensor(
+            ch_obj.propagation_dir or [0.0, 1.0, 0.0], dtype=torch.float64
+        )
+        prop_dir /= prop_dir.norm()
+
+        # each channel has a noise on its laser amplitude
+        # we assume each channel has the same noise amplitude currently
+        # the hardware currently has only a global channel anyway
+        sigma_factor = (
+            1.0
+            if amp_sigma == 0.0
+            else torch.max(torch.tensor(0), torch.normal(1.0, amp_sigma, (1,))).item()
+        )
+        for slot in ch_samples.slots:
+            factors = (
+                torch.tensor(
+                    [
+                        math.exp(-((perp_dist(x, prop_dir) / laser_waist) ** 2))
+                        for x in qubit_positions
+                    ],
+                    dtype=torch.float64,
+                )  # the lasers have a gaussian profile perpendicular to the propagation direction
+                if laser_waist and ch_obj.addressing == "Global"
+                else torch.ones(
+                    len(q_ids), dtype=torch.float64
+                )  # but for a local channel, this does not matter
+            )
+
+            # add the amplitude noise for the targeted qubits
+            factors[[x in slot.targets for x in q_ids]] *= sigma_factor
+
+            for i in range(slot.ti, slot.tf):
+                if i in times_to_amp_factors:  # multiple local channels at the same time
+                    # pulser enforces that no two lasers target the same qubit simultaneously
+                    # so only a single factor will be != 1.0 for each qubit
+                    times_to_amp_factors[i] = factors * times_to_amp_factors[i]
+                else:
+                    times_to_amp_factors[i] = factors
+    return times_to_amp_factors
+
+
+def _get_delta_offset(nqubits: int, temperature: float) -> torch.Tensor:
+    """
+    The delta values are shifted due to atomic velocities.
+    The atomic velocities follow the Maxwell distribution
+    https://en.wikipedia.org/wiki/Maxwell%E2%80%93Boltzmann_distribution
+    and then a given residual velocity is converted to a delta offset per
+    https://en.wikipedia.org/wiki/Doppler_broadening
+    """
+    if temperature == 0.0:
+        return torch.zeros(nqubits, dtype=torch.float64)
+    t = temperature * 1e-6  # microKelvin -> Kelvin
+    sigma = KEFF * math.sqrt(KB * t / RUBIDIUM_MASS)
```

Comment:
I'm a bit surprised with these small numbers. Dont you think that it's more safe to introduce a ration `KEFF*sqrt(KB/RUBIDIUM_MASS)` instead of separate entities

Reply:
I merged `KB` and `RUBIDIUM_MASS` into `KB_PER_RUBIDIUM_MASS`, and it doesn't appreciably change anything. I've kept the change because I think it's alright, but I think merging it further would destroy understanding of where the constants come from, without adding numerical benefit.